hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java

```diff
@@ -83,27 +87,34 @@ private void invokeOnAllPartitions() {
             Address address = mp.getKey();
             List<Integer> partitions = mp.getValue();
             PartitionIteratingOperation operation = new PartitionIteratingOperation(operationFactory, partitions);
-            Future future = operationService.createInvocationBuilder(serviceName, operation, address)
+
+            InternalCompletableFuture future = operationService.createInvocationBuilder(serviceName, operation, address)
                     .setTryCount(TRY_COUNT)
                     .setTryPauseMillis(TRY_PAUSE_MILLIS)
                     .invoke();
             futures.put(address, future);
         }
     }
 
-    private void awaitCompletion() {
+    private void awaitCompletion() throws Exception {
         NodeEngineImpl nodeEngine = operationService.nodeEngine;
-        for (Map.Entry<Address, Future> response : futures.entrySet()) {
+        for (Map.Entry<Address, InternalCompletableFuture> response : futures.entrySet()) {
             try {
-                Future future = response.getValue();
-                PartitionResponse result = (PartitionResponse) nodeEngine.toObject(future.get());
+                InternalCompletableFuture future = response.getValue();
+                Object o = nodeEngine.toObject(future.get());
+                PartitionResponse result = (PartitionResponse) o;
                 result.addResults(partitionResults);
-            } catch (Throwable t) {
+            } catch (Exception t) {
+                if (!isRetryable(t)) {
```

Comment:
You can even create just a single operation (not invoked) when you encounter an exception and feed it with the exception just here, in catch block. 
If it reports that exception is retryable then continue, otherwise throw the exception. That single operation can be reused whenever a new exception is caught.